dartagnan/src/main/java/com/dat3m/dartagnan/program/processing/MemToReg.java

```diff
@@ -0,0 +1,353 @@
+                if (registers != null) {
+                    replacingRegisters.put(allocation, registers.stream().map(function::newRegister).toList());
+                    assert allocation.getUsers().isEmpty();
+                    allocation.tryDelete();
+                }
+            }
+        }
+        int loadCount = 0, storeCount = 0;
+        // Replace all loads and stores to replaceable storage.
+        for (final Map.Entry<MemoryEvent, AddressOffset> entry : matcher.accesses.entrySet()) {
+            final MemoryEvent event = entry.getKey();
+            final AddressOffset access = entry.getValue();
+            final List<Register> registers = access == null ? null : replacingRegisters.get(access.base);
+            if (registers == null || access.offset < 0 || access.offset >= registers.size()) {
+                continue;
+            }
+            if (event instanceof Load load) {
+                final Register reg = load.getResultRegister();
+                assert load.getUsers().isEmpty();
+                load.replaceBy(EventFactory.newLocal(reg, expressions.makeCast(registers.get(access.offset), reg.getType())));
+                loadCount++;
+            }
+            if (event instanceof Store store) {
+                final Register reg = registers.get(access.offset);
+                assert store.getUsers().isEmpty();
+                store.replaceBy(EventFactory.newLocal(reg, expressions.makeCast(store.getMemValue(), reg.getType())));
+                storeCount++;
+            }
+        }
+        // Remove involved local assignments.
+        for (final Map.Entry<Local, AddressOffset> entry : matcher.assignments.entrySet()) {
+            if (replacingRegisters.containsKey(entry.getValue().base)) {
+                assert entry.getKey().getUsers().isEmpty();
+                entry.getKey().tryDelete();
+            }
+        }
+        if (loadCount + storeCount > 0) {
+            logger.debug("Removed {} loads and {} stores in function \"{}\".", loadCount, storeCount, function.getName());
+        }
+    }
+
+    private List<Type> getReplacingRegisters(Alloc allocation) {
+        if (!(allocation.getArraySize() instanceof IValue sizeExpression)) {
+            return null;
+        }
+        final int size = sizeExpression.getValueAsInt();
+        if (size != 1) {
+            //TODO arrays
+            return null;
+        }
+        final List<Type> replacementTypes = new ArrayList<>();
+        final Type type = allocation.getAllocationType();
+        //TODO PointerType
+        if (type instanceof IntegerType || type instanceof BooleanType) {
+            replacementTypes.add(type);
+        } else {
+            //TODO aggregate types
+            return null;
+        }
+        //TODO check for mixed-size accesses
+        return replacementTypes;
+    }
+
+    // Invariant: base != null
+    private record AddressOffset(RegWriter base, int offset) {
```

Comment:
You could also state the invariant with `@Nonnull` annotations on the record component.